File: day17.lua
io.input("data/input17.txt")
local inp=io.read("*all")
local A,B,C,pr=string.match(inp,"Register A: (%d+)\nRegister B: (%d+)\nRegister C: (%d+)\n\nProgram: (.+)")
A,B,C=tonumber(A),tonumber(B),tonumber(C)
local prog={}
for a in string.gmatch(pr,"(%d+)") do
    table.insert(prog,tonumber(a))
end

function execute(A,B,C)
    local ip,output=0,0
    while ip<#prog do
        local op1,op2=prog[ip+1],prog[ip+2] --because of 1-based indexing
        local comb
        if op2<=3 then comb=op2
        elseif op2==4 then comb=A
        elseif op2==5 then comb=B
        elseif op2==6 then comb=C end
        if op1==0 then A=A//(1<<comb) end
        if op1==1 then B=op2~B end
        if op1==2 then B=comb%8 end
        if op1==3 and A~=0 then ip=op2-2 end
        if op1==4 then B=B~C end
        if op1==5 then output=output*10+comb%8 end
        if op1==6 then B=A//(1<<comb) end
        if op1==7 then C=A//(1<<comb) end
        ip=ip+2
    end
    return output
end

local part1=execute(A,B,C)

function searchDigit(pos,start,step)
    if pos<0 then return start end
    for i=0,7 do
        local A,B,C=start+i*step,0,0
        local output=execute(A,B,C)
        local digit=tonumber(string.sub(tostring(output),pos+1,pos+1))
        if digit==prog[pos+1] then
            local s=searchDigit(pos-1,start+i*step,step//8)
            if s then return s end
        end
    end
end
local start=1<<3*(#prog-1)
local blocksize=start
local part2=searchDigit(#prog-1,start,blocksize)
print(part1,part2)

File: day17.py
with open("data/input17.txt") as f:
    reg,prog=f.read().strip().split("\n\n")
reg=reg.split()
reg=[int(reg[2]),int(reg[5]),int(reg[8])]
prog=[int(k) for k in prog[9:].split(",")]

def combo(n,reg):
    if 0<=n<=3:
        return n
    if 4<=n<=6:
        return reg[n-4]
    if n==7:
        print("erreur")

def execute(prog,reg):
    output=[]
    ip=0
    while ip<len(prog):
        #print("...reg", reg)
        op1,op2=prog[ip],prog[ip+1]
        comb=combo(op2,reg)
        if op1==0:
            num,denom=reg[0],2**comb
            reg[0]=num//denom
            ip+=2
        if op1==1:
            reg[1]=op2^reg[1]
            ip+=2
        if op1==2:
            reg[1]=comb%8
            ip+=2
        if op1==3:
            if reg[0]!=0:
                ip=op2
            else:
                ip+=2
        if op1==4:
            reg[1]=reg[1]^reg[2]
            ip+=2
        if op1==5:
            output.append(comb%8)
            ip+=2
        if op1==6:
            num,denom=reg[0],2**comb
            reg[1]=num//denom
            ip+=2
        if op1==7:
            num,denom=reg[0],2**comb
            reg[2]=num//denom
            ip+=2
    return output

part1=execute(prog,reg)

def searchDigit(digitpos,start, stepsize):
    if digitpos<0:
        return start
    for i in range(8):
        reg[0]=start+i*stepsize
        reg[1]=0
        reg[2]=0
        output=execute(prog,reg)
        if output[digitpos]==prog[digitpos]:
            s=searchDigit(digitpos-1,start+i*stepsize,stepsize//8)
            if s: return s
 
start=8**(len(prog)-1)
blocksize=8**(len(prog)-1)
part2=searchDigit(len(prog)-1,start,blocksize)

print(part1,part2)